guides/c.md

```diff
@@ -0,0 +1,169 @@
+Introduction
+============
+
+The C wrappers should look like the C++ code as much as possible. But, there are going to be differences between the wrappers and the C++ code for many reasons. When in doubt, refer to to the wrapped library's style guides, or make a best guess based on code already written in either the wrapped library, or other wrappers.
+
+Formatting
+==========
+
+Use autoformatting where possible. Generally the wrapped libraries use clang-format, so it's best to use that. Also, it is suggested to do formatting on save.
+
+Implementation
+==============
+
+- If there are function overrides, then prefer to wrap the core implementation. If there's multiple implementations, then try to pick a suffix that is clear and understandable.
+
+Error Handling
+==============
+
+TODO
+
+SUGGESTION: Implement something like Rust's `Result<T, E>`, unless the wrapped library already implements error handling that we can inherit from in C.
+
+```c
+enum StatusCode {
+    OK;
+    ERROR;
+};
+typedef struct {
+    StatusCode status;
+    union value {
+        MyType value;
+        Error err;
+    }
+} MyTypeStatus;
+```
+
+Exceptions
+==========
+
+C does not support exceptions, and it's generally considered bad to allow exceptions to pass across the C++ to C boundary (TODO: Why?). So, if wrapped code can cause an exception, then it should be wrapped in code like below.
+
+TODO: This may not catch everything.
+
+```cpp
+ThingStatus mything_do_thing(int thing) {
+    ThingStatus status;
+
+    try {
+        Thing thing = mything::do_thing(int thing);
+
+        status.status = StatusCode.OK;
+        status.value = thing;
+
+        return status
+    } catch (const std::exception& err) {
+        status.status = StatusCode.ERROR;
+        status.err = err;
+        return status;
+    }
+};
+```
+
+Static Asserts
+==============
+
+It is critical to make sure that the types passed through the C++ to C interface are always exactly the same if the type is non-opaque. So, when writing the bindings, there must be a static assert to validate that the C types are the same as the C++ types.
+
+```cpp
+// C++ class being wrapped
+namespace mything {
+class Thing {
+    int a;
+    int b;
+    float c;
+};
+}
+
+extern "C" {
+// C wrapper
+typedef struct {
```

Comment:
I still have some opinions about this.
My view is:
* If the type is visible to c with no public fields you can just use a char array for it's size + alignment attribute.
* If it's opaque with no public fields, then you don't even need a struct declaration. You are just passing pointers around, so you can just forward declare it.